Make batched Google API calls more efficient

nbey opened the pull request:
- [x] Group calls for all attachments
- [x] Use batch methods for both teacher and student attached drive files
- [x] Implement Logger
- [x] Split batched requests into groups of < 100

nbey committed:
Make batched Google API calls more efficient

- Group calls for all attachments
- Use batch methods for both teacher and student attached drive files

nbey committed:
Retry failed Google API requests

nbey committed:
Extend request timeout limit when saving tasks

themightychris commented on php-classes/Google/API.php:
```diff
@@ -104,7 +112,7 @@ public static function executeBatchRequest(array $Requests)
                     $requestKey, // 2
                     $Request->getMethod(), // 3
-                    $Request->getUri()->getPath() . ($Request->getUri()->getQuery() ? '?'.$Request->getUri()->getQuery() : ''), // 4
+                    $Request->getUri()->getPath() . "?quotaUser=$requestKey". ($Request->getUri()->getQuery() ? '&'.$Request->getUri()->getQuery() : ''), // 4
```
store the result of getQuery

themightychris commented on php-classes/Google/API.php:
```diff
@@ -141,17 +151,39 @@ public static function executeBatchRequest(array $Requests)
         }
 
+        if (!empty($failedRequests)) {
+            \Emergence\Mailer\Mailer::send('nafis@jarv.us', "Failed Responses (#$retry)", join('', ['<pre>', print_r($failedResponses, true), '</pre>']));
```
Use the general logger instead, and in config files on your site you can change what log levels get emailed and where they get emailed

themightychris reviewed (changes requested):
minor tweaks, mostly just implement logging like you indicated in the TODO checklist and get rid of your hardcoded email address and mailing code

nbey committed:
Store request query as a variable

nbey committed:
Improve batch requests workflow

- Implement logging, defaulting to `Emergence\Logger` instances.
- Extract some logic into private/protected methods for clarity
- Verified failed requests have a 'valid' reason for retrying
- Split requests > 100 into groups
- Reorder use statements
